src/kudu/consensus/routing.cc

```diff
@@ -0,0 +1,519 @@
+    DCHECK(result.second) << "unexpected duplicate uuid: " << node_uuid;
+  }
+
+  *index = std::move(tmp_index);
+  *forest = std::move(tmp_forest);
+
+  // This is just a warning, not an error.
+  if (!proxy_from_nodes_not_found.empty()) {
+    return Status::Incomplete(
+        "the following proxy_from nodes specified in the proxy topology were "
+        "not found in the active Raft config and have been ignored",
+        JoinStrings(proxy_from_nodes_not_found, ", "));
+  }
+
+  return Status::OK();
+}
+
+Status RoutingTable::MergeForestIntoSingleRoutingTree(
+    const std::string& leader_uuid,
+    const std::unordered_map<std::string, Node*>& index,
+    std::unordered_map<std::string, std::unique_ptr<Node>>* forest) {
+  Node* leader = FindWithDefault(index, leader_uuid, nullptr);
+  if (!leader) {
+    return Status::InvalidArgument("invalid config: cannot find leader",
+                                   leader_uuid);
+  }
+
+  // Find the ultimate proxy root of the leader, if the leader as a proxy
+  // assigned to it.
+  Node* source_root = leader;
+  while (source_root->proxy_from) {
+    source_root = source_root->proxy_from;
+  }
+
+  // Make all trees, except the one the leader is in, children of the leader.
+  // The result is a single tree.
+  auto iter = forest->begin();
+  while (iter != forest->end()) {
+    if (iter->first == source_root->id()) {
+      ++iter;
+      continue;
+    }
+    const string& child_uuid = iter->first;
+    iter->second->proxy_from = leader;
+    leader->children.emplace(child_uuid, std::move(iter->second));
+    iter = forest->erase(iter);
+  }
+
+  DCHECK_EQ(1, forest->size());
+  return Status::OK();
+}
+
+void RoutingTable::ConstructNextHopIndicesRec(Node* cur) {
+  for (const auto& child_entry : cur->children) {
+    const string& child_uuid = child_entry.first;
+    const auto& child = child_entry.second;
+    ConstructNextHopIndicesRec(child.get());
+    // Absorb child routes.
+    for (const auto& child_route : child->routes) {
+      const string& dest_uuid = child_route.first;
+      cur->routes.emplace(dest_uuid, child_uuid);
+    }
+  }
+  // Add self-route as a base case.
+  cur->routes.emplace(cur->id(), cur->id());
+}
+
+Status RoutingTable::NextHop(const string& src_uuid,
+                             const string& dest_uuid,
+                             string* next_hop) const {
+  Node* src = FindWithDefault(index_, src_uuid, nullptr);
+  if (!src) {
+    return Status::NotFound(Substitute("unknown source uuid: $0", src_uuid));
+  }
+  Node* dest = FindWithDefault(index_, dest_uuid, nullptr);
+  if (!dest) {
+    return Status::NotFound(Substitute("unknown destination uuid: $0", dest_uuid));
+  }
+
+  // Search children.
+  string* next_uuid = FindOrNull(src->routes, dest_uuid);
+  if (next_uuid) {
+    *next_hop = *next_uuid;
+    return Status::OK();
+  }
+
+  // If we can't route via a child, route via a parent.
+  DCHECK(src->proxy_from);
```

Comment:
this DCHECK might bomb in prod if there is a new instance that has joined the config,
or its a pending config change. The routing topology might not have been Updated yet.

The other way to say it is what happens if the the routing Table was built (Inited) and then a new
peer made it to the config.

Reply:
This is impossible assuming we don't have a thread safety bug because the tree is always connected. Therefore any route is either down or up. DurableRoutingTable uses a read-write-commit lock to provide thread safety for an encapsulated instance of RoutingTable.